### DISCUSSION TITLE
"Matter Casting" (a.k.a. MatterCast) audio player (streaming reciever) for new/upcoming video and music cast open standard for local streaming?

### DISCUSSION THREAD

**Discussion by Hedda** [music-assistant/support#2342] [Feature requests and ideas]
Please consider researching and try to make a plan to add a custom "**_Matter Casting_**" (a.k.a. MatterCast) player provider with receiver/client componets for Music Assistant in the future, and then later consider also adding a "Matter Casting" streaming service as server-side for streaming to upcoming connected smart-speakers and smart-displays/televisions using this ”Matter Casting” standard so that Music Assistent clients be able to act as receiver endpoints and audio/music player for "Matter Casting" audio/music streaming (and video streaming) when those become available. The main issue looks to be that the Matter Casting specification seems to currently be missing support for the audio-only part of its architecture as their casting technology focus so far has just been on video streaming:

- https://github.com/project-chip/connectedhomeip/issues/31389

But still, this new ”Matter Casting” standard is aimed at democratizing local video and audio casting in a universal way that can be supported by all ecosystems and platforms, and most importantly have a great chance of being adopted by all major tech companies.

* https://www.matteralpha.com/explainer/what-is-matter-casting (Note! unofficial explainer blog)

* https://matter-smarthome.de/en/know-how/what-is-matter-casting/ (Note! another unofficial blog)

In sunmery; "**Matter Casting**" is a new open protocol media streaming standard and the Matter Casting APIs for casting video and audio streams over a local network is only a small part of the currently much-hyped Matter standard suite for IoT which is being led, promoted and developed by the [CSA (Connectivity Standards Alliance) and its very impressive list of member companies](https://csa-iot.org/members/):

* [https://github.com/project-chip/connectedhomeip/blob/a4300e7f3a8999e5b283aeae20e39d645b296a67/examples/tv-casting-app/APIs.md](https://github.com/project-chip/connectedhomeip/blob/master/examples/tv-casting-app/APIs.md)

* [https://github.com/project-chip/connectedhomeip/tree/master/examples/tv-casting-app](https://github.com/project-chip/connectedhomeip/tree/master/examples/tv-casting-app)

This is simply meant as an informal feature request that can also be as an open discussion about this new open video and audio streaming standard and Matter Casting API/protocol that might eventually become an industry-standard in the future, meant to replace or compete directly with proprietary and closed source media streaming protocols used by [Google Cast](https://en.wikipedia.org/wiki/Google_Cast) (Google Chromecast), Apple's [AirPlay (formerly AirTunes for AirPort)](https://en.wikipedia.org/wiki/AirPlay), Alexa Cast (Amazon Alexa / Amazon Echo), and [Miracast](https://en.wikipedia.org/wiki/Miracast). Being an enthusiast of open-source software and non-proprietary standards I am hoping for industry-wide adoption of this new open standard for casting video and audio streams in the future.

Anyway, "**Matter Casting**" (also wrongly referred to as "Matter Cast") is a technology standard that uses open APIs to enable you to cast video and/or audio content from your smartphone, tablet, and computer to your TV or smart display and smart speakers / audio-recievers + voice assistant appliances. From an end-user's point-of-view Matter Casting is meant to work the same thing as Google Cast, Apple AirPlay, and Miracast or other similar streaming media protocols for sharing multimedia on a local network, with a few major differences: implementing support for this "Matter Casting" is royalty-free and it is based on open-source code, so not proprietary. That means it is designed from the ground up to be a "free and open source" video and audio casting/streaming standard for the best possible cross-platform compatibility. 

Matter casting is an open protocol that does not require specific hardware, only that the app on your source device and the app on the target device you are casting to both have the feature enabled. So again, from an end-users point-of-view it is essentially like Apple’s AirPlay or Google Cast but can be available to every app or hardware maker to implement and not limited to specific mobile OS and ecosystems, or partnerships, however under the hood Matter Casting is also superior to those other protocols in many ways.

https://www.techhive.com/article/2200884/matter-casting-looks-superior-to-airplay-chromecast-can-it-win.html

"***In a brief demo at CES, Amazon showed how users would get all the features of its Prime Video app while casting, including its X-Ray feature, and could then back out from the video player to the Prime Video menu. Chris DeCenzo, the senior principal engineer at Amazon who led Matter Casting development, confirmed that with linear apps such as Sling TV, you might be able to flip through channels with the TV remote after launching a channel through Matter Casting.***

***“It actually is a different architecture from what you see with the other casting protocols,” DeCenzo says. “Matter Casting is really about communicating with apps on the TV.”***

***Chromecast doesn’t work quite the same way. On Android TV and Google TV devices, Chromecast’s video streams are usually separate from the in-app versions, so you can’t get full controls with your regular remote. While Google does offer [a way for apps to connect directly with Chromecast](https://go.redirectingat.com/?id=111346X1569484&url=https://developers.googleblog.com/2020/08/introducing-cast-connect-android-tv.html&sref=https://www.techhive.com/article/2200884/matter-casting-looks-superior-to-airplay-chromecast-can-it-win.html&xcust=3-1-2200884-1-0-0), I haven’t seen any major streaming services use this feature except Netflix. And if you’re using Chromecast on a different TV platform such as Vizio SmartCast, there’s no app connection at all.***

***Apple’s AirPlay has a similar limitation: Videos you launch via AirPlay are separate from the apps on your TV, and using AirPlay even precludes you from playing other media on the iPhone or iPad from which the video is streaming.***

***All of which means that Matter Casting will feel more like an extension of the apps on your TV than an entirely separate way to watch video. That might make you more likely to use it in the first place.***"

It is under early development and still looks to be missing real-time audio synchronization to allow for synchronized multi-room music playback, but at least is it already supporting Android, Darwin (for iOS and Mac), and Linux, allowing any developer to implement it into its products. Hopefully ensuring wider-spread compatibility between devices of different manufacturers and ecosystems. 

So the reason why I think this is interesting is that it is the first attempt by the technology consortium governing the overall [Matter standard suite](https://en.wikipedia.org/wiki/Matter_(standard)) to create also an open standard for local video and audio casting that is meant for broader adoption across different ecosystems for interoperability. 

[https://csa-iot.org/all-solutions/matter/](https://csa-iot.org/all-solutions/matter/)

Because the Matter standard matters if adopted as the foundation for all connected things:

[https://csa-iot.org/newsroom/building-a-standard-that-really-matters/](https://csa-iot.org/newsroom/building-a-standard-that-really-matters/)

Matter in general already has hundreds of member companies, but understand that not everyone has publicly committed to every part of that standard, as most have only announced support for the IoT parts that control smart devices, however, [support for additional device types is being announced regularly](https://csa-iot.org/newsroom/matter-1-2-arrives-with-nine-new-device-types-improvements-across-the-board). "**Matter Casting**" is needed in this toolkit for streaming audio and video locally in a unified way.

Matter Casting adoption has so far been announced by Amazon who will first add support to newer Amazon Fire TV and Amazon Echo Show series of smart display devices via an upcoming firmware update, (in which it sounds like they might initially only implement video streaming?).

More interestingly, Amazon has already embraced Matter Casting fully have publicly stated that they do not plan on ever adding support for Google Cast or Apple AirPlay video casting to their newer Fire OS based devices. This is just hot from the press of the CES 2024 (Consumer Electronics Show yearly tech event):

https://www.aboutamazon.com/news/devices/amazon-ces-2024-announcements

"_**Amazon is a long-time supporter of using open technology standards to give customers more choice over the devices and services they use in their homes. That’s also why Amazon is a founding member and key contributor to the Matter Standard. At CES, Amazon announced Matter Casting, which enables customers to cast content to Fire TV and Echo Show 15 devices directly from supported streaming apps on iOS and Android. Customers can begin watching a movie or browse for their next favorite show from Prime Video on their phone, and cast it to their compatible Fire TV device or Echo Show 15. This is an industry-first demonstration of implementing Matter Casting.**_"

https://www.theverge.com/2024/1/9/24030324/amazon-matter-casting-echo-show-fire-tv-prime-video

"_**Chris DeCenzo (a principal engineer at Amazon who works on Alexa and Fire OS devices) says that if all the big companies adopted Matter Casting, it would be better for the industry, as app developers would only have to develop for one protocol — one of the founding principles of the Matter smart home standard. “We believe in open standards,” he says, “because it simplifies things for developers.**_"

"_**Additionally, it would make things easier for customers. Casting from apps would work with every streaming box, TV, and smart display, and you wouldn’t have to think about what hardware you have in order to cast your content. That sounds great in theory, but — as with Matter in general — it looks like it will be a long journey to get there.**_"

https://www.aboutamazon.com/news/devices/amazon-ces-2024-announcements

**Comment by seyfeb** [comment]
Does MatterCast allow synchronized streaming of video to, e.g., a screen and audio to MatterCast speakers?

**Reply by Hedda** [reply]
There is not yet Matter Casting multiroom support or even audio-only streaming in the upstream framework far as I understand but Amazon developers and project manager(s) working on this new "Matter Casting" streaming protocol have not revealed a roadmap or future goals so it is unclear, see upstream feature request and discussion here:

- https://github.com/project-chip/connectedhomeip/issues/31389

Also note there that are not yet any Matter Casting speakers since it is so far only focus on video casting and not audio-only casting for music, etc. without video. So audio streaming is technically only supported if part of a video stream. Please see related downsteam feature request for Home Assistant's developers here that is specifically requesting those two features:

- https://community.home-assistant.io/t/matter-casting-matter-casting-client-support-in-home-assistant-cast-new-upcoming-open-protocol-standard-for-local-video-and-audio-streaming/671645

**Comment by Hedda** [comment]
FYI, [chrisdecenzo](https://github.com/chrisdecenzo) (Chris DeCenzo from Amazon who is the lead Matter Casting developer and a member is in the Matter working group, a.k.a. Project CHIP) wrote today that there is now an official effort within the Matter working group (in Project CHIP) right now to define use cases for "Matter Casting" to audio players and smart speakers or voice assistants within the Matter standard (but only Matter members can join a Matter working group and participate with that work as far as known), see his comment here about joining that effort -> https://github.com/project-chip/connectedhomeip/issues/31389#issuecomment-2172863098

@marcelveldt As you are both the founder of Music Assistant and "the Matter guy" at Nabu Casa (as well as the lead Matter integration developer for Home Assistant) who is a member of Matter and the CSA/CSA-IOT perhaps this is something that you (and maybe also ESPHome developers) at Nabu Casa would be interested in helping architecture and develop for the Matter project? 

* https://www.nabucasa.com/about/
* https://www.home-assistant.io/blog/2024/05/09/music-assistant-2

I would think that epecially audio-only streaming with focus on high-quality audio-streams and music-playback + multi-room support functionality might be very relative to current Home Assistant roadmap now that Home Assistant's recent announcementabout both your "Music Assistant" add-on and the Open Home Foundation + the Home Assistant's voice assistant work, which are separate things that I believe all align in spirit with this concept at a high-level?


* https://www.home-assistant.io/blog/2024/06/12/roadmap-2024h1/
  * https://www.home-assistant.io/blog/2024/06/12/roadmap-introduction/

PS: I also read that Nabu Casa is developing your own [ESPHome](https://github.com/esphome) based smart-speaker and voice-assistant hardware, and for such type of devices to also work as music streaming end-point and HiFi audio-player (A/V-receiver endpoint) with your "Music Assistant" integration acting as middleware I am guessing you are eventually going to want to add support for some cross-ecosystem support for some kind of standardized audio streaming protocol like Matter Casting with audio player endpoint features?

* https://github.com/esphome/feature-requests/issues/2723

**Comment by Hedda** [comment]
FYI, they updated Matter Casting APIs documentation in `tv-casting-app` as part of v1.3 Matter APIs:

* https://github.com/project-chip/connectedhomeip/blob/7ee93d2bd4519859bd773ff66b92695d6aece317/examples/tv-casting-app/APIs.md

They also updated their `tv-app` (Matter TV Casting app example):

* https://github.com/project-chip/connectedhomeip/tree/master/examples/tv-app

#### *Enhanced Entertainment and Smart Home Management Functionality for Media Devices*

"***Matter Casting Media Players / TVs** — Matter 1.3 offers improvements to TV functionality, including push messages and dialog support for new ambient experiences, casting initialization enhancements, expanded interactivity options for TV apps, text and track support, and improved search functionality. Interaction with other devices in the home is also enhanced, enabling other Matter devices to send notifications to TVs or other devices with screens (for instance, notifying that a robot vacuum is stuck, the laundry is done, etc.)*"

https://csa-iot.org/newsroom/matter-1-3-specification-released/

**Comment by Hedda** [comment]
FYI, The Verge has got an update from Chris LaPré, CTO of the Connectivity Standards Alliance (CSA) and posted a new article about Matter Casting, most notible saying that “a new streaming speaker device type and related controls” is under development a new Matter complatible smart home standard for smart speakers is being led by a former Sonos executive looking to disrupt the smart speaker market, (the article even mentions Home Assistant though how exactly it fits into this context at this time is unclear):

* https://www.theverge.com/2025/1/2/24332848/speakers-matter-smart-home-apple-google-amazon-legato

*According to Fiede Schillmoeller, CEO of [Legato](https://www.legatoxp.com/), the Dutch company is leading the group developing this new Matter smart speaker specification.*

*Legato is working with CSA member companies that are speaker manufacturers, software makers, ecosystems, and content providers, said Schillmoeller. The group is “fully focused on making audio and speakers work through Matter,” he said. While he declined to name names, he said, “It includes everyone you would want in that group.” (It’s worth noting that both Sonos and Bose are [members of the CSA](https://csa-iot.org/members/participants/).)*

*“We want smart homes to have this ambient information stream [and] speakers are a big enabler of that”*

*The specifics of how speakers will work are still under development, so Schillmoeller couldn’t share exact details, but he confirmed that the team is building something that will handle “all the important stuff.”*

*This should include choosing your audio source, controlling volume and full transport controls, “play, pause, skip ... all these things that help you control the speaker from any device you like,” he said. If implemented, this would mean you could control any Matter speaker from any Matter-supported ecosystem app, such as Apple Home, Alexa, Google Home, and **Home Assistant**.*

*Schillmoeller also hopes the team can integrate speakers with your smart home through scenes. “You should be able to use your smart home to drive speaker behavior,” he said. “Set what type of music plays in the morning when you wake up, what happens to your speaker when you leave the house, what happens to the audio when someone rings your doorbell.” Functions like these are available in smart home ecosystems today but are limited to proprietary software and hardware.*

*Matter won’t enable multiroom music, a feature likely to remain at the ecosystem level*.

*What Matter probably won’t do is enable multiroom music, a feature likely to remain at the ecosystem level. Additionally, don’t hold out hope that Apple and Google, and possibly Amazon, will enable their speakers as Matter device types. While it’s technically possible with Matter, controlling your Apple HomePod from your Google Nest Hub just doesn’t feel likely. I’d love to be proven wrong, though.*

**Comment by Hedda** [comment]
FYI, not quite sure if it is indirectly related to this or not but just spotted that @lboue has open two draft pull requests against Home Assistant's core that will implement a "Matter Speaker" device using the MediaPlayer entity, so not if such a "Matter nedia player platform" relates to this requested “Matter Casting” (“Matter Casting Client” feature in any way or if it only define the control-plane for using Matter audio control cluster from a remote control device to allow it turn a Matter speaker ON and OFF or changing the volume + muting for Matter speakers?

- https://github.com/home-assistant/core/pull/150470

  - https://github.com/home-assistant/core/pull/150104
